### PR TITLE
Fix issue with folding of CASE/IIF

### LINKED ISSUE #49387: SQL: "Should not fold expression" for conditional function
`SELECT IIF(20 = 0, NULL, salary / 20) FROM test_emp`

results in

```
org.elasticsearch.xpack.sql.SqlIllegalArgumentException: Should not fold expression
        at org.elasticsearch.xpack.sql.expression.Expression.fold(Expression.java:72) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.predicate.BinaryPredicate.fold(BinaryPredicate.java:32) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.predicate.conditional.Case.fold(Case.java:132) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.Literal.of(Literal.java:159) ~[?:?]
        at org.elasticsearch.xpack.sql.expression.Literal.of(Literal.java:144) ~[?:?]
        at org.elasticsearch.xpack.sql.optimizer.Optimizer$ConstantFolding.rule(Optimizer.java:1238) ~[?:?]
        at org.elasticsearch.xpack.sql.tree.Node.transformDown(Node.java:174) ~[?:?]
```

### PR DESCRIPTION
**Pull request by matriv**
Add extra checks to prevent ConstantFolding rule to try to fold
the CASE/IIF functions early before the SimplifyCase rule gets applied.

Fixes: #49387

### CODE REVIEW

**Review by astefan — APPROVED** [review]
LGTM

**Review by costin — CHANGES_REQUESTED** [review]
I consider it a semantic leak when expression classes become aware of runtime rules.
It indicates that either the expression tries to do too much (in this case folding) or that the relationship between the rules is not correct (either the rules need to be simplified or the order between them changed).

The comment refers to a rule but it seems to me that the folding is eager?
Or maybe `SimplifyCase` be moved before `ConstantFolding` - I couldn't tell from the test case what is the current issue...

**Comment by matriv** [comment]
@costin The issue was that the Case was answering `foldable() -> true` even if the result of the condition (folded to true) was un-foldable. The tests test that the case is not foldable before the SimplifyCase rule takes place.

I thought of changing the order but I thought personally that can become easier a future confusion than my current solution. The current code just tests that the result or elseResult (depending on the TRUE/FALSE of the condition) is also foldable and (at least in my mind) doesn't appear that connected to the rules and their order. On the contrary the previous state seems to depend more on the ordering of the rules in the Optimizer, and the comment on the method (comes from a previous suggestion of yours) clearly states that it's based on the fact that the SimplifyCase rule has been applied.

If you think changing the orders of the rules in the Optimizer seems a better approach, I'm happy to proceed with that.

**Comment by matriv** [comment]
@costin 
To give an example for supporting the current solution:
```
public void testSimplifyIif_ConditionFalse_NonFoldableResult() {
        SimplifyCase rule = new SimplifyCase();
        Iif iif = new Iif(EMPTY, new Equals(EMPTY, ONE, TWO), Literal.of(EMPTY, "foo"), getFieldAttribute("myField"));
        assertFalse(iif.foldable()); --> previously here it was TRUE which logically is wrong.

        Expression e = rule.rule(iif);
        assertEquals(Iif.class, e.getClass());
        iif = (Iif) e;
        assertEquals(0, iif.conditions().size());
        assertFalse(iif.foldable());
        assertEquals("myField", Expressions.name(iif.elseResult()));
    }
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Case.java [hunk 1]
```diff
@@ -120,8 +120,18 @@ protected TypeResolution resolveType() {
      */
     @Override
     public boolean foldable() {
-        return (conditions.isEmpty() && elseResult.foldable()) ||
-            (conditions.size() == 1 && conditions.get(0).condition().foldable() && conditions.get(0).result().foldable());
+        if (conditions.isEmpty() && elseResult.foldable()) {
+            return true;
+        }
+        if (conditions.size() == 1 && conditions.get(0).condition().foldable()) {
+            // Need to prevent ConstantFolding rule to get applied before SimplifyCase rule
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Case.java]: I find the comment misleading as it implies the rules order is incorrect - called in a different order the rules might affect `fold` but the matter is, `foldable` is incorrect (which leads to the exception).

**Review by costin — APPROVED** [review]
LGTM - the comment about the rules threw me off hence why it should be removed.

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/default-distro
@elasticmachine run elasticsearch-ci/bwc

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Pull request merged**